src/components/clusterexplorer/explorer.ts

```diff
@@ -0,0 +1,123 @@
+import { ClusterExplorerNode, ClusterExplorerResourceNode } from './node';
+import { MiniKubeContextNode, ContextNode } from './node.context';
+
+export const KUBERNETES_EXPLORER_NODE_CATEGORY = 'kubernetes-explorer-node';
+
+export type KubernetesExplorerNodeType =
+    'error' |
+    'context' |
+    'folder.resource' |
+    'folder.grouping' |
+    'resource' |
+    'configitem' |
+    'helm.release' |
+    'extension';
+
+export function create(kubectl: Kubectl, host: Host): KubernetesExplorer {
+    return new KubernetesExplorer(kubectl, host);
+}
+
+export function isKubernetesExplorerResourceNode(obj: any): obj is ClusterExplorerResourceNode {
+    return obj && obj.nodeCategory === KUBERNETES_EXPLORER_NODE_CATEGORY && obj.nodeType === 'resource';
+}
+
+export class KubernetesExplorer implements vscode.TreeDataProvider<ClusterExplorerNode> {
+    private onDidChangeTreeDataEmitter: vscode.EventEmitter<ClusterExplorerNode | undefined> = new vscode.EventEmitter<ClusterExplorerNode | undefined>();
+    readonly onDidChangeTreeData: vscode.Event<ClusterExplorerNode | undefined> = this.onDidChangeTreeDataEmitter.event;
+
+    private readonly extenders = Array.of<ExplorerExtender<ClusterExplorerNode>>();
+    private readonly customisers = Array.of<ExplorerUICustomizer<ClusterExplorerNode>>();
+
+    constructor(private readonly kubectl: Kubectl, private readonly host: Host) {
+        host.onDidChangeConfiguration((change) => {
+            if (affectsUs(change)) {
+                this.refresh();
+            }
+        });
+    }
+
+    getTreeItem(element: ClusterExplorerNode): vscode.TreeItem | Thenable<vscode.TreeItem> {
+        const treeItem = element.getTreeItem();
+
+        const treeItem2 = providerResult.transform(treeItem, (ti) => {
+            if (ti.collapsibleState === vscode.TreeItemCollapsibleState.None && this.extenders.some((e) => e.contributesChildren(element))) {
+                ti.collapsibleState = vscode.TreeItemCollapsibleState.Collapsed;
+            }
+        });
+
+        let treeItem3 = treeItem2;
+        for (const c of this.customisers) {
+            treeItem3 = providerResult.transformPossiblyAsync(treeItem2, (ti) => c.customize(element, ti));
+        }
+        return treeItem3;
+}
+
+    getChildren(parent?: ClusterExplorerNode): vscode.ProviderResult<ClusterExplorerNode[]> {
+        const baseChildren = this.getChildrenBase(parent);
+        const contributedChildren = this.extenders
+                                        .filter((e) => e.contributesChildren(parent))
+                                        .map((e) => e.getChildren(parent));
+        return providerResult.append(baseChildren, ...contributedChildren);
+    }
+
+    private getChildrenBase(parent?: ClusterExplorerNode): vscode.ProviderResult<ClusterExplorerNode[]> {
+        if (parent) {
+            return parent.getChildren(this.kubectl, this.host);
+        }
+        return this.getClusters();
+    }
+
+    refresh(): void {
```

Comment:
you don't need an explicit `void` in TypeScript do you?

Reply:
If I don't put the `void` then TypeScript infers the return type.  Specifying `void` catches the case where a maintainer inadvertently changes it to return something - I want them to be forced to document to the return type if that happens.